bfdd/bfd.c

```diff
@@ -558,6 +568,10 @@ void ptm_bfd_sess_dn(struct bfd_session *bfd, uint8_t diag)
 				   state_list[bfd->ses_state].str,
 				   get_diag_str(bfd->local_diag));
 	}
+
+	/* clear peer's mac address */
+	UNSET_FLAG(bfd->flags, BFD_SESS_FLAG_MAC_SET);
+	memset(bfd->peer_hw_addr, 0, sizeof(bfd->peer_hw_addr));
```

Comment:
Please also reset the the local address otherwise if we change the local address the new generated echo packets will have the previous address.

```c
memset(&bfd->local_address, 0, sizeof(bfd->local_address));
```

Example:
```
interface foo
 ip address 192.168.0.10/24
!
bfd
 peer 192.168.0.20
  echo
 !
!
```

After connection is up and echo has been seen, change the address:

```
interface foo
 no ip address 192.168.0.10/24
 ip address 192.168.0.11/24
!
```

When the BFD peer is up and echo starts the packet loop will fail because the source/destination address will be wrong (using the old one).

I haven't tried a setup with multiple addresses and changing between them, but I think it is ok to drop the connection to end up cleaning it here (the source address changed too after all).